nsg-netbox.py

```diff
@@ -51,96 +101,331 @@ def run(self):
-            filters[fk] = fv
-        if self.config.get('whitelist'):
+        if self.config and self.config.get('filters'):
+            for fk, fv in self.config['filters'].items():
+                if 'custom_fields' in fk:
+                    fk = 'cf_{}'.format(fk.split('.')[1])
+                filters[fk] = fv
+        else:
+            filters = {"status": self.active}
+
+        if self.config and self.config.get('whitelist'):
             filters['tag'] = self.config['whitelist']
-        return nbox.dcim.devices.filter(**filters)
 
-    def get_primary_ip(self, device):
-        return str(ipaddress.ip_interface(device.primary_ip).ip)
+        devices = self.netbox_dcim(model="devices", filters=filters)
+        for device in devices:
+            ip = self.get_primary_ip(device)
+            if ip:
+                result[ip] = device
+            else:
+                host = f"{device.name}.{self.domain}" if self.domain and self.domain not in device.name else device.name
+                try:
+                    ip = socket.gethostbyname(host)
+                    if ip:
+                        result[ip] = device
+                    else:
+                        self.log.error("get_netbox_devices: can't resolve device name: {}".format(device.name))
+                except Exception as e:
+                    self.log.error("get_netbox_devices: can't resolve device name: {}: {}".format(device.name, e))
+            site = self._sites.get(device.site.name)
+            if isinstance(site, pynetbox.core.response.Record):
+                for key in ('facility', 'continent', 'metro', 'cluster'):
+                    if not hasattr(device, key) or not getattr(device, key):
+                        setattr(device, key, getattr(site, key) if hasattr(site, key) else None)
+            device.nsg_tags = self.extract_tags(device)
+        return result
+        # return {self.get_primary_ip(d): d for d in self.netbox_dcim(nbox) if self.condition(d)}
 
-    def condition(self, device):
-        if self.config.get('blacklist'):
+    def get_netbox_sites(self):
+        sites = self.netbox_dcim(model="sites")
+        for site in sites:
+            metro = site.name[:3]
+            try:
+                if not hasattr(site, "facility"):
+                    setattr(site, "facility", None)
+                if not hasattr(site, "cluster"):
+                    setattr(site, "cluster", None)
+                for tag in site.tags:
+                    if tag.name.lower() in ("facility", "pop"):
+                        site.facility = site.name
+                        site.cluster = None
+                    elif tag.name.lower() in ("cluster",):
+                        site.facility = site.name[:-1]
+                        site.cluster = site.name
+                setattr(site, "continent", self.get_mapping(config_key="metro_to_continent", key=metro))
+                setattr(site, "metro", metro)
```

Comment:
this program should not try to generate tags. If a tag is not in NetBox, we do not "create" it. This means we do not need to convert metro to continent.